src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataCollectionSerializer.cs

```diff
@@ -90,6 +91,12 @@ public ODataCollectionSerializer(ODataSerializerProvider serializerProvider)
                 {
                     collectionStart.NextPageLink = writeContext.InternalRequest.Context.NextLink;
                 }
+                else if (writeContext.InternalRequest.Context.QueryOptions != null)
+                {
+                    // Collection serializer is called only for collection of primitive values - A null object will be supplied since it is a non-entity value
+                    SkipTokenHandler skipTokenHandler = writeContext.QueryOptions.Context.GetSkipTokenHandler();
+                    collectionStart.NextPageLink = skipTokenHandler.GenerateNextPageLink(writeContext.InternalRequest.RequestUri, writeContext.InternalRequest.Context.PageSize, null, writeContext);
```

Comment:
>null [](start = 177, length = 4)

shouldn't we pass the item anyway, and let the skiptokenhandler decide if it can do the paging or not? In the future, if the service supports [ordered collections](https://github.com/oasis-tcs/odata-vocabularies/blob/master/vocabularies/Org.OData.Core.V1.xml#L401) of primitive/complex types, they could do paging over those ordered collections as long as we provide them the value.

Reply:
not fixed in iteration 5

---
In reply to: [291339533](https://github.com/OData/WebApi/pull/1724#discussion_r291339533) [](ancestors = 291339533)

Reply:
As I had discussed with you that the WriteStart writes the next link which is problematic in getting the last object without traversing through the collection again. It is documented here - [https://github.com/OData/odata.net/issues/1486](https://github.com/OData/odata.net/issues/1486)

---
In reply to: [293533405](https://github.com/OData/WebApi/pull/1724#discussion_r293533405) [](ancestors = 293533405,291339533)